Build failure with gcc 11.x on native_posix

**Describe the bug**
Warnings from new gcc turning into errors:

```
In file included from /home/nashif/Work/zephyrproject/zephyr/include/kernel_includes.h:39,
                 from /home/nashif/Work/zephyrproject/zephyr/include/kernel.h:17,
                 from /home/nashif/Work/zephyrproject/zephyr/include/zephyr.h:18,
                 from /home/nashif/Work/zephyrproject/zephyr/kernel/init.c:14:
/home/nashif/Work/zephyrproject/zephyr/include/sys/thread_stack.h:155:16: error: ignoring attribute ‘section (".noinit.\"WEST_TOPDIR/zephyr/kernel/init.c\".3")’ because it conflicts with previous ‘section (".noinit.\"WEST_TOPDIR/zephyr/kernel/include/kernel_internal.h\".0")’ [-Werror=attributes]
  155 |         struct z_thread_stack_element __kstackmem \
      |                ^~~~~~~~~~~~~~~~~~~~~~
/home/nashif/Work/zephyrproject/zephyr/kernel/init.c:61:1: note: in expansion of macro ‘K_KERNEL_STACK_ARRAY_DEFINE’
   61 | K_KERNEL_STACK_ARRAY_DEFINE(z_interrupt_stacks, CONFIG_MP_NUM_CPUS,
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/nashif/Work/zephyrproject/zephyr/kernel/include/kernel_internal.h:136:36: note: previous declaration here
  136 | extern K_KERNEL_STACK_ARRAY_DEFINE(z_interrupt_stacks, CONFIG_MP_NUM_CPUS,
      |                                    ^~~~~~~~~~~~~~~~~~
/home/nashif/Work/zephyrproject/zephyr/include/sys/thread_stack.h:157:17: note: in definition of macro ‘K_KERNEL_STACK_ARRAY_DEFINE’
  157 |                 sym[nmemb][Z_KERNEL_STACK_LEN(size)]
      |                 ^~~
cc1: all warnings being treated as errors
gmake[2]: *** [zephyr/kernel/CMakeFiles/kernel.dir/build.make:134: zephyr/kernel/CMakeFiles/kernel.dir/init.c.obj] Error
1
gmake[1]: *** [CMakeFiles/Makefile2:2781: zephyr/kernel/CMakeFiles/kernel.dir/all] Error 2
gmake: *** [Makefile:103: all] Error 2

```

**To Reproduce**

On fedora 34 with gcc 11.x

Steps to reproduce the behavior:
1.  twister -T samples/philosophers/ -p native_posix
4. See error

--- a/include/arch/x86/arch.h
+++ b/include/arch/x86/arch.h
@@ -154,7 +154,7 @@ static ALWAYS_INLINE uint32_t sys_read32(mm_reg_t addr)
 static ALWAYS_INLINE void sys_set_bit(mem_addr_t addr, unsigned int bit)
 {
 	__asm__ volatile("btsl %1, %0"
-			 : "+m" (*(volatile uint32_t *) (addr))
+			 : "+m" (*(volatile uint8_t *) (addr))
 			 : "Ir" (bit)
 			 : "memory");
 }
@@ -162,7 +162,7 @@ static ALWAYS_INLINE void sys_set_bit(mem_addr_t addr, unsigned int bit)
 static ALWAYS_INLINE void sys_clear_bit(mem_addr_t addr, unsigned int bit)
 {
 	__asm__ volatile("btrl %1, %0"
-			 : "+m" (*(volatile uint32_t *) (addr))
+			 : "+m" (*(volatile uint8_t *) (addr))
 			 : "Ir" (bit));
 }
 
@@ -172,7 +172,7 @@ static ALWAYS_INLINE int sys_test_bit(mem_addr_t addr, unsigned int bit)
 
 	__asm__ volatile("btl %2, %1;"
 			 "sbb %0, %0"
-			 : "=r" (ret), "+m" (*(volatile uint32_t *) (addr))
+			 : "=r" (ret), "+m" (*(volatile uint8_t *) (addr))
 			 : "Ir" (bit));
 
 	return ret;
@@ -185,7 +185,7 @@ static ALWAYS_INLINE int sys_test_and_set_bit(mem_addr_t addr,
 
 	__asm__ volatile("btsl %2, %1;"
 			 "sbb %0, %0"
-			 : "=r" (ret), "+m" (*(volatile uint32_t *) (addr))
+			 : "=r" (ret), "+m" (*(volatile uint8_t *) (addr))
 			 : "Ir" (bit));
 
 	return ret;
@@ -198,7 +198,7 @@ static ALWAYS_INLINE int sys_test_and_clear_bit(mem_addr_t addr,
 
 	__asm__ volatile("btrl %2, %1;"
 			 "sbb %0, %0"
-			 : "=r" (ret), "+m" (*(volatile uint32_t *) (addr))
+			 : "=r" (ret), "+m" (*(volatile uint8_t *) (addr))
 			 : "Ir" (bit));
 
 	return ret;

--- a/include/kernel/thread_stack.h
+++ b/include/kernel/thread_stack.h
@@ -123,6 +123,36 @@ static inline char *z_stack_ptr_align(char *ptr)
  */
 
 /**
+ * @def K_KERNEL_STACK_ARRAY_EXTERN
+ * @brief Obtain an extern reference to a stack array
+ *
+ * This macro properly brings the symbol of a stack array declared
+ * elsewhere into scope.
+ *
+ * @param sym Thread stack symbol name
+ * @param nmemb Number of stacks to declare
+ * @param size Size of the stack memory region
+ */
+#define K_KERNEL_STACK_ARRAY_EXTERN(sym, nmemb, size) \
+	extern struct z_thread_stack_element \
+		sym[nmemb][Z_KERNEL_STACK_LEN(size)]
+
+/**
+ * @def K_KERNEL_PINNED_STACK_ARRAY_EXTERN
+ * @brief Obtain an extern reference to a pinned stack array
+ *
+ * This macro properly brings the symbol of a pinned stack array
+ * declared elsewhere into scope.
+ *
+ * @param sym Thread stack symbol name
+ * @param nmemb Number of stacks to declare
+ * @param size Size of the stack memory region
+ */
+#define K_KERNEL_PINNED_STACK_ARRAY_EXTERN(sym, nmemb, size) \
+	extern struct z_thread_stack_element \
+		sym[nmemb][Z_KERNEL_STACK_LEN(size)]
+
+/**
  * @def Z_KERNEL_STACK_DEFINE_IN
  * @brief Define a toplevel kernel stack memory region in specified section
  *
@@ -272,6 +302,7 @@ static inline char *Z_KERNEL_STACK_BUFFER(k_thread_stack_t *sym)
 #define K_THREAD_STACK_MEMBER		K_KERNEL_STACK_MEMBER
 #define Z_THREAD_STACK_BUFFER		Z_KERNEL_STACK_BUFFER
 #define K_THREAD_STACK_EXTERN		K_KERNEL_STACK_EXTERN
+#define K_THREAD_STACK_ARRAY_EXTERN	K_KERNEL_STACK_ARRAY_EXTERN
 #else
 /**
  * @def K_THREAD_STACK_RESERVED
@@ -370,6 +401,20 @@ static inline char *Z_KERNEL_STACK_BUFFER(k_thread_stack_t *sym)
  * @param sym Thread stack symbol name
  */
 #define K_THREAD_STACK_EXTERN(sym) extern k_thread_stack_t sym[]
+
+/**
+ * @brief Obtain an extern reference to a thread stack array
+ *
+ * This macro properly brings the symbol of a stack array declared
+ * elsewhere into scope.
+ *
+ * @param sym Thread stack symbol name
+ * @param nmemb Number of stacks to declare
+ * @param size Size of the stack memory region
+ */
+#define K_THREAD_STACK_ARRAY_EXTERN(sym, nmemb, size) \
+	extern struct z_thread_stack_element \
+		sym[nmemb][K_THREAD_STACK_LEN(size)]
 
 /**
  * @addtogroup thread_stack_api

--- a/include/kernel/thread_stack.h
+++ b/include/kernel/thread_stack.h
@@ -104,6 +104,9 @@ static inline char *z_stack_ptr_align(char *ptr)
 #define Z_KERNEL_STACK_OBJ_ALIGN	ARCH_STACK_PTR_ALIGN
 #endif
 
+#define Z_KERNEL_STACK_LEN(size) \
+	ROUND_UP(Z_KERNEL_STACK_SIZE_ADJUST(size), Z_KERNEL_STACK_OBJ_ALIGN)
+
 /**
  * @brief Obtain an extern reference to a stack
  *
@@ -203,9 +206,6 @@ static inline char *z_stack_ptr_align(char *ptr)
 #define K_KERNEL_PINNED_STACK_DEFINE(sym, size) \
 	Z_KERNEL_STACK_DEFINE_IN(sym, size, __kstackmem)
 #endif
-
-#define Z_KERNEL_STACK_LEN(size) \
-	ROUND_UP(Z_KERNEL_STACK_SIZE_ADJUST(size), Z_KERNEL_STACK_OBJ_ALIGN)
 
 /**
  * @def K_KERNEL_STACK_ARRAY_DEFINE

--- a/kernel/include/kernel_internal.h
+++ b/kernel/include/kernel_internal.h
@@ -152,7 +152,7 @@ extern struct k_thread z_main_thread;
 #ifdef CONFIG_MULTITHREADING
 extern struct k_thread z_idle_threads[CONFIG_MP_NUM_CPUS];
 #endif
-extern K_KERNEL_STACK_ARRAY_DEFINE(z_interrupt_stacks, CONFIG_MP_NUM_CPUS,
+K_KERNEL_PINNED_STACK_ARRAY_EXTERN(z_interrupt_stacks, CONFIG_MP_NUM_CPUS,
 				   CONFIG_ISR_STACK_SIZE);
 
 #ifdef CONFIG_GEN_PRIV_STACKS

--- a/tests/kernel/common/src/bitarray.c
+++ b/tests/kernel/common/src/bitarray.c
@@ -11,7 +11,6 @@
 #include <tc_util.h>
 #include <sys/bitarray.h>
 #include <sys/util.h>
-#include <toolchain.h>
 
 #ifdef CONFIG_BIG_ENDIAN
 #define BIT_INDEX(bit)  ((3 - ((bit >> 3) & 0x3)) + 4*(bit >> 5))
@@ -338,13 +337,29 @@ void alloc_and_free_predefined(void)
 		     "sys_bitarray_free() failed bits comparison");
 }
 
+static inline size_t count_bits(uint32_t val)
+{
+	/* Implements Brian Kernighan’s Algorithm
+	 * to count bits.
+	 */
+
+	size_t cnt = 0;
+
+	while (val != 0) {
+		val = val & (val - 1);
+		cnt++;
+	}
+
+	return cnt;
+}
+
 size_t get_bitarray_popcnt(sys_bitarray_t *ba)
 {
 	size_t popcnt = 0;
 	unsigned int idx;
 
 	for (idx = 0; idx < ba->num_bundles; idx++) {
-		popcnt += popcount(ba->bundles[idx]);
+		popcnt += count_bits(ba->bundles[idx]);
 	}
 
 	return popcnt;

--- a/tests/kernel/mem_protect/mem_protect/src/kobject.c
+++ b/tests/kernel/mem_protect/mem_protect/src/kobject.c
@@ -362,7 +362,7 @@ void test_kobject_release_from_user(void)
  */
 void test_kobject_release_null(void)
 {
-	int dummy;
+	int dummy = 0;
 
 	k_object_release(&dummy);
 }

--- a/tests/kernel/sched/schedule_api/src/test_sched.h
+++ b/tests/kernel/sched/schedule_api/src/test_sched.h
@@ -14,7 +14,7 @@
 #define STACK_SIZE (640 + CONFIG_TEST_EXTRA_STACKSIZE)
 
 K_THREAD_STACK_EXTERN(tstack);
-extern K_THREAD_STACK_ARRAY_DEFINE(tstacks, MAX_NUM_THREAD, STACK_SIZE);
+K_THREAD_STACK_ARRAY_EXTERN(tstacks, MAX_NUM_THREAD, STACK_SIZE);
 extern struct k_thread user_thread;
 extern struct k_sem user_sem;
 

